### ISSUE TITLE
Container runtime reads partially written CNI plugin config and fails

### ISSUE THREAD

**Issue by fawadkhaliq** [aws/amazon-vpc-cni-k8s#1437]
**What happened**
Container runtime can read a partially written CNI config and fail. This is because ipamd installation/restart writes the CNI plugin configuration directly to [CNI config directory](https://github.com/aws/amazon-vpc-cni-k8s/blob/master/scripts/entrypoint.sh#L109) (e.g. `/host/etc/cni/net.d`). The container runtime net config sync loop (e.g. in [containerd](https://github.com/containerd/containerd/blob/0867c99b1a69cda79ae7354cbe54b05ffb6093ac/pkg/cri/server/cni_conf_syncer.go#L73)) gets triggered on fs change events and if the contents of CNI config are not fully written, container runtime will fail with an error like `unexpected end of JSON input`. 

We should write to tmp and move the _fully written_ file to CNI config directory

**Attach logs**
```
msg="failed to reload cni configuration after receiving fs
change event(\"/etc/cni/net.d/10-aws.conflist\": WRITE)" error="cni config load failed: failed to load CNI config list file /etc/cni/net.d/10-aws.conflist: error parsing configuration li
st: unexpected end of JSON input: invalid cni config: failed to load cni config"
```

**What you expected to happen**
We should write the conf to a temp file and move the already written file to CNI config directory so the container runtime always sees a fully written file and not incomplete/invalid JSON forcing an unnecessary failure on the container runtime. 

**How to reproduce it (as minimally and precisely as possible)**:
It's a race. Restart ipamd and container runtime at the same time.

**Anything else we need to know?**
Nope

**Environment**
- Kubernetes version (use `kubectl version`): N/A
- CNI Version: All VPC CNI versions
- OS (e.g: `cat /etc/os-release`): N/A
- Kernel (e.g. `uname -a`): N/A

**Comment by ZF-fredericvanlinthoudt** [comment]
@fawadkhaliq
Correct me if I'm wrong but I think this issue should get a higher priority.
Since this morning we saw this issue happening at 3 EKS clusters (v1.20) and for all it started around the same time (maybe due to some network problems in our region, I don't know but the 3 clusters are in 3 different AWS accounts so it is a coincidence it started around the same time).

What I've observed is that on certain nodes the aws-node pod was crash looping, while other nodes were running successfully.
I did cordon & drain the problematic nodes and then terminated them manually so they get replaced by fresh instances in the autoscaling group.
When the new instances were online, aws-node worked fine again, but then 1 other node, that previously did not have the issue, suddenly also started to have problems.
So I then executed the same actions (cordon, drain, terminate) as before on that node, and with a fresh instance everything was ok again.

We did not perform any changes on the 3 EKS clusters and yet we faced the issue, probably because ipamd has restarted for some reason within the aws-node container?
If this happens on production clusters, this is a real pain because from then on no new containers can be started on the problematic nodes because the aws-node pod is crash looping and they do not get the correct networking to get started.
